Avoid re-lexing, move position to internal state

gagern opened the pull request:
Instead of passing around the current position as an argument, we now have a parser property called `pos` to keep track of that.  Instead of repeatedly re-lexing at the current position we now have a property called `nextToken` which contains the token beginning at the current position.  We may need to
re-lex if we switch mode.  Since the position is kept in the parser state, we don't need to return it from parsing methods, which obsoletes the `ParseResult` class.

This file contains two hacks (search for `TODO` in the diff) aimed at keeping error messages exactly as they were. Once #385 gets merged, we can improve these locations and adjust the error message expectations accordingly. Right now, this PR won't conflict with #385 but doesn't depend on it either. In the long run, we might even go one step further and switch from reporting an error *position* to reporting an error *token* or *group*. We could underline these ad a whole.

This is the last step in addressing #266. It prepares the way for implementing macros.

kevinbarabash commented:
Thanks for reminding me about this diff.  I will try to look at it this evening.  Sorry about the delay.

gagern commented:
I'm glad you don't mind being reminded this way. :grinning:

kevinbarabash commented on src/Parser.js:
```diff
@@ -70,13 +69,21 @@ function ParseFuncOrArgument(result, isFunction) {
  */
-Parser.prototype.expect = function(result, text) {
-    if (result.text !== text) {
+Parser.prototype.expect = function(text, consume) {
```
Could update the docstring to include a description of `consume`?

kevinbarabash commented on src/Parser.js:
```diff
@@ -70,13 +69,21 @@ function ParseFuncOrArgument(result, isFunction) {
+    }
+};
+
+Parser.prototype.consume = function() {
```
Could you add brief description of this function?

kevinbarabash commented on src/Parser.js:
```diff
@@ -87,18 +94,20 @@ Parser.prototype.expect = function(result, text) {
 Parser.prototype.parse = function(input) {
```
Could you get rid of `input`?

kevinbarabash commented on src/Parser.js:
```diff
@@ -141,15 +150,15 @@ Parser.prototype.parseExpression = function(pos, breakOnInfix, breakOnToken) {
-    var res = new ParseResult(this.handleInfixNodes(body), pos);
-    res.peek = lex;
-    return res;
+    return this.handleInfixNodes(body);
```
Looking at this code again, I realize that I could've done this in a better way.

kevinbarabash commented on src/Parser.js:
```diff
@@ -211,31 +220,30 @@ var SUPSUB_GREEDINESS = 1;
         } else {
             throw new ParseError(
-                "Expected group after '" + symbol + "'", this.lexer, pos);
+                "Expected group after '" + symbol + "'", this.lexer, this.pos);
```
Could you change `this.pos` to `symPos`?

kevinbarabash commented on src/parseData.js:
```diff
@@ -7,17 +7,7 @@ function ParseNode(type, value, mode) {
-function ParseResult(result, newPosition, peek) {
-    this.result = result;
-    this.position = newPosition;
-}
```
Yay!

kevinbarabash commented on src/Parser.js:
```diff
@@ -278,76 +288,62 @@ Parser.prototype.handleSupSubscript = function(pos, symbol, name) {
  *
  * @return {?ParseResult}
```
{?ParseNode}

kevinbarabash commented on src/Parser.js:
```diff
@@ -392,52 +386,47 @@ var styleFuncs = [
+    var start = this.parseSymbol();
 
-    if (!start || !start.result) {
+    if (start === null) {
```
Can we use `start == null` instead?

gagern replied:
Why? Most projects I work with prefer `===` over `==` for its cleaner semantics and its better performance.

kevinbarabash replied:
At KA we prefer using `==` (or `!=`) when comparing against `null` because it will produce the same results if `start` is `undefined`.  In this case it's pretty easy to verify that `start` is either a `ParseNode` or `null` so it's not a big deal, but in other cases it may not be easy to make this determination.

gagern replied:
OK, I changed this occurrence. Using `==` for `null` only is certainly better than always using it. For a moment I feared I might have to change many other strict comparisons as well.

kevinbarabash commented on src/Parser.js:
```diff
@@ -488,12 +474,14 @@ Parser.prototype.parseImplicitGroup = function(pos) {
+Parser.prototype.parseFunction = function(baseGroup) {
+    if (!baseGroup) {
+        baseGroup = this.parseGroup();
+    }
```
It feels a little weird having this as an optional parameter, but I don't have a good suggestion of how to make it less weird.  Could you update the jsdocs to reflect that `baseGroup` is optional?

kevinbarabash commented on src/Parser.js:
```diff
@@ -534,77 +518,73 @@ Parser.prototype.callFunction = function(name, args, positions) {
     var totalArgs = funcData.numArgs + funcData.numOptionalArgs;
     if (totalArgs === 0) {
-        return pos;
+        return [[this.pos]];
```
If the return is an array of arguments with the last element being an array of positions, I would expect that we should be returning `[[]]` here.

kevinbarabash replied:
I hope in the future that the args can contain position data so that we don't need this extra element in the array.

gagern replied:
No, `positions` is [defined](https://github.com/gagern/KaTeX/blob/21a26b807cb1c3b411257e9eddd581a7fe48b2b6/src/functions.js#L68-L69) as

> the positions in the overall string of the function and the arguments.

So even without arguments, the position of the function itself is required. This is in line with how `positions` is [initialized](https://github.com/gagern/KaTeX/blob/21a26b807cb1c3b411257e9eddd581a7fe48b2b6/src/Parser.js#L548) in the case of a non-zero argument count. And in any case, this is preserving existing functionality, we shouldn't change anything here as part of the new position handling.

Moving positions into e.g. parse nodes and the likes will be greatly simplified with all these changes here, I think.

kevinbarabash replied:
I forgot about that, thanks.

kevinbarabash commented on src/Parser.js:
```diff
@@ -624,43 +604,46 @@ Parser.prototype.parseSpecialGroup = function(pos, innerMode, optional) {
         }
-        var closeBrace = this.lexer.lex(inner.position, outerMode);
-        this.expect(closeBrace, optional ? "]" : "}");
+        this.consume();
```
Could you add a comment about what this is consuming?

kevinbarabash commented on src/Parser.js:
```diff
@@ -624,43 +604,46 @@ Parser.prototype.parseSpecialGroup = function(pos, innerMode, optional) {
-        var whitespace = this.lexer.lex(pos, "whitespace");
-        pos = whitespace.position;
+        var whitespace = this.lexer.lex(this.pos, "whitespace");
+        this.pos = whitespace.position;
```
Could probably make `_innerLex` always ignore leading whitespace and then we wouldn't need to this special casing.  This is something that should probably go in a separate commit or PR.

gagern replied:
What's your definition of “leading” here? Right now I don't see how this could be possible while still getting correct spacing inside `\text{…}` constructs.

kevinbarabash replied:
Sorry I confused whitespace before `\text{}` with whitespace within `\text{}` before non-whitespace text within `\text{}`.

kevinbarabash commented on src/Parser.js:
```diff
@@ -670,23 +653,21 @@ Parser.prototype.parseSpecialGroup = function(pos, innerMode, optional) {
+Parser.prototype.parseGroup = function() {
+    var start = this.nextToken;
     // Try to parse an open brace
     if (start.text === "{") {
```
Could just test against `this.nextToken.text` directly instead of introducing `start`

kevinbarabash commented on src/Parser.js:
```diff
@@ -695,19 +676,17 @@ Parser.prototype.parseGroup = function(pos) {
+Parser.prototype.parseOptionalGroup = function() {
+    var start = this.nextToken;
     // Try to parse an open bracket
     if (start.text === "[") {
```
Same comment as in `parseGroup`

kevinbarabash commented:
Looks pretty good.  I'll do another pass of the weekend just to make sure I'm not missing anything.  I'm glad we have the tests that we do otherwise I'd be much less confident about accepting pull requests that make changes like this one.

gagern commented:
I adressed most of your concerns; those I didn't comment on. Also fixed one off-by-one error in one error message position. If this looks fine to you, let me know and I'll squash and rebase it.

kevinbarabash commented:
@gagern the changes look good for the most part.  I made a couple of comments on the most recent commit.  After those are address please squash and rebase.

gagern commented:
Fixed and squashed. The docstring for the `consume` parameter now reads like this:

```
 * @param {boolean=} consume whether to consume the expected token,
 *                           defaults to true
```

kevinbarabash commented:
Looks good.  Thanks for the quick response.
